### ISSUE TITLE
acceptance: Core Stability Testing Scenarios

### ISSUE THREAD

**Issue by BramGruneir** [cockroachdb/cockroach#4015]
This issue is being setup to track a collection of tests that will ensure that we've covered a subset of representative scenarios with a variety of cluster setups. 
Note that this list is not designed to be exhaustive but to have broad coverage over a collection of different scenarios, each to designed to push the db to the limit.  And with each of these scenarios, we need to test them with a variety of test cluster setups, listed below.

Scenarios to run:
1. Rebalancing - apply an unequal load and then an unequal unload
2. Contention - both SSI and SI
3. Parallelism - multiple requests that don't contend with each other, overload the system
4. Correctness - bank example? or something else
5. Deletion - full load the db, than remove almost all of it
6. Overfill - do we handle an overfilled db correctly
7. Faulty Hardware - node and/or store flakiness or outright failure
8. Slow - node and/or store
9. Catastrophe and Recover - lose all or almost all nodes and bring them all back up one at a time
10. Catastrophe and Recover - lose all or almost all nodes and bring them all back up one at the same time
11. Partition - cut the cluster in two, let both halves run independently briefly, then join them back up

Test cluster setups:
The format is node count with store count, so 1:1 is one node with a single store, 3:2,1:1 would be 3 nodes with 2 stores each and 1 node with a single store.
Note that this doesn't cover stores of different sizes, perhaps we should randomize store sizes.  Also, testing with a small number of nodes with multiple stores won't be possible until #2067 is addressed.
1. Single Node
   - 1:1
   - 1:2
   - 1:3
   - 1:10
2. Dual Node
   - 2:1
   - 2:2
3. Tri-Node
   - 3:1
   - 1:1, 1:2, 1:3
4. Higher Node Systems
   - 4:1
   - 5:1
   - 1:1, 1:2, 1:3, 1:4, 1:5
   - 7:1
   - 10:1
   - 5:1, 5:2
   - 20:1

Variables to watch in all tests:
- Latency/throughput
- Availability
- Frequency of leadership changes 
- Rebalancing events
- Consistency - perhaps once the scenario is complete and the cluster is stable

Once we've agreed on a final list of testing scenarios and cluster setups, I'll turn this into a checklist with issues for each scenario/setup so we can check them off individually.

Please comment on this list!  Suggest other scenarios or perhaps redundant scenarios and setups.

**Comment by bdarnell** [comment]
Multiple stores per node are not very interesting with fewer than 4 nodes due to #2067. We should do a little testing with these configurations to make sure things don't break horribly, but most of our testing with 1-3 nodes should only use one store until that issue is addressed.

For catastrophe/recovery tests, we should include both cases where nodes start back up one at a time and cases where they all come back at once. We also need to test partitions (nodes are still up but can't talk to each other; make sure they can catch up and don't overload peers with pent-up messages)

Variables to watch in all tests:
* Latency/throughput
* Availability
* Frequency of leadership changes and rebalancing events

**Comment by petermattis** [comment]
Not a testing scenario per-se, but worth capturing here: we need either offline or online cluster consistency checks. For example, after a test runs and the cluster is quiescent we should verify that all of the replicas for a range have identical data. At a higher level, we should verify the consistency of SQL table data and associated indexes.

**Comment by BramGruneir** [comment]
thanks @bdarnell, incorporated all your feedback.

@petermattis, I agree, we discussed adding this check yesterday but I was thinking it should be called out in its own issue.  I'll add one shortly.  And I added consistency in the Variables to Watch that @bdarnell suggested.

**Comment by bdarnell** [comment]
Consistency checker is #837

**Comment by BramGruneir** [comment]
Great.  I think it's something we should start working on in the new week or two.

**Comment by vivekmenezes** [comment]
I'm thinking of adding an acceptance test similar to the banking example that will allow us to bring up a cluster and run the bank example against it and ensure that the performance variables we care about are in good stead, all while running chaos monkey in the background.

That will be followed up by adding more complex schema and load generation scenarios. We also expect to have a way of configuring the test using a config file with a variety of knobs which will allow us to run many instances of the same test with different setups.

**Comment by petermattis** [comment]
@BramGruneir Is this issue still useful to keep open?

**Comment by BramGruneir** [comment]
I think we're good to close it.
